pkg/sidecars/server/server.go

```diff
@@ -0,0 +1,117 @@
+package sidecarserv
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/inspr/inspr/pkg/sidecars/models"
+	"go.uber.org/zap"
+)
+
+// Server is a struct that contains the variables necessary
+// to handle the necessary routes of the rest API
+type Server struct {
+	broker       string
+	Reader       models.Reader
+	Writer       models.Writer
+	inAddr       string
+	outAddr      string
+	client       *http.Client
+	runningRead  bool
+	runningWrite bool
+}
+
+var logger *zap.Logger
+
+func init() {
+	logger = zap.NewNop()
+	// logger, _ = zap.NewProduction(zap.Fields(zap.String("section", "loadbalencer-sidecar")))
+}
+
+// Init - configures the server
+func Init(r models.Reader, w models.Writer, broker string) *Server {
+	server := &Server{}
+	// server fetches addresses variable names from models.
+	envVars := models.GetSidecarConnectionVars(broker)
+	if envVars == nil {
+		panic(fmt.Sprintf("%s broker's enviroment variables not configured", broker))
+	}
+	server.broker = broker
+
+	// server fetches required addresses from deployment.
+	inAddr, ok := os.LookupEnv(envVars.WriteEnvVar)
+	if !ok {
+		panic(fmt.Sprintf("[ENV VAR] %s not found", envVars.WriteEnvVar))
+	}
+
+	outAddr, ok := os.LookupEnv(envVars.ReadEnvVar)
+	if !ok {
+		panic(fmt.Sprintf("[ENV VAR] %s not found", envVars.ReadEnvVar))
+	}
+
+	server.inAddr = fmt.Sprintf(":%s", inAddr)
+	server.outAddr = fmt.Sprintf("http://localhost:%v", outAddr)
+	server.client = &http.Client{}
+
+	// implementations of write and read for a specific sidecar
+	server.Reader = r
+	server.Writer = w
+	return server
+} // looked
+
+// Run starts the server on the port given in addr
+func (s *Server) Run(ctx context.Context) error {
+	server := &http.Server{
+		Handler: s.writeMessageHandler().Post().JSON(), // look writeMessageHandler - OK
+		Addr:    s.inAddr,
+	}
+	errCh := make(chan error)
+	// create read message routine and captures its error
+	go func() { errCh <- s.readMessageRoutine(ctx) }() // readMessageRoutine
+
+	var err error
+	go func() {
+		s.runningWrite = true
+		defer func() { s.runningWrite = false }()
+		if err = server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			logger.Error("listen:%v", zap.Error(err))
+		}
+	}()
+
+	logger.Info("sideCar listener is up...")
+
+	select {
+	case <-ctx.Done():
+		s.Writer.Close()
```

Comment:
It's probably a good idea to move these `s.Writer.Close()` in the `select` to inside `gracefulShutdown(...)` to avoid unecessary code replication